### PR TITLE
[Dynamo] De-dup graph inputs

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98845
* __->__ #98775

###  Overview
This PR de-duplicates graph inputs in TorchDynamo, using the `Source` as the unique identifier for each input. This closes https://github.com/pytorch/pytorch/issues/98743 and https://github.com/pytorch/pytorch/issues/98625.

### Details
`VariableBuilder.wrap_tensor()` should return a `VariableTracker` for the passed-in `value: Tensor`. If `value` is duplicated, we should avoid calling `OutputGraph.create_graph_input()` and `OutputGraph.add_grapharg()`.
- Note that `create_graph_input()` and `add_grapharg()` are not 1:1. For a constant source and either `wrap_sym()` or `wrap_unspecialized_primitive()`, TorchDynamo still calls `create_graph_input()` but not `add_grapharg()`.
- Note that `create_graph_input()` should be called before constructing the corresponding `VariableTracker`. TorchDynamo needs the `fx.Proxy` object to pass to `wrap_fx_proxy()`. 

In this PR, the `OutputGraph` saves an additional mapping `input_source_to_var` from each graph input's `Source` to its `VariableTracker`, which works because `Source` is now hashable. This mapping should be updated each time `create_graph_input()` is called. However, since we must construct the `VariableTracker` after `create_graph_input()` returns, we must have a separate call to the `OutputGraph` to update the mapping.

If anyone has any suggestion on how to coalesce this logic and avoid having to remember to update `input_source_to_var` for each `create_graph_input()`, I would love to hear it.

<details>
<summary> Alternate Approach</summary>

Initially, I tried having TorchDynamo construct a new but equivalent `VariableTracker` for the duplicated tensor. However, I abandoned this approach after hitting an assertion in `def wrap_fx_proxy_cls()` due to `"example_value"` already being in the proxy node's metadata because we were reusing the primary tensor's `Proxy` object. Reusing the exact `VariableTracker` also seems less error-prone instead of requiring constructing a new but identical `VariableTracker`.
</details>

#### Global Variable Test
```
import torch
@torch.compile()
def f(): 
    return x + x
x = torch.randn(3)
f()
```

Before:
```
====== Forward graph 0 ======
 <eval_with_key>.6 class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: f32[3], arg1_1: f32[3]):
        # File: /data/users/ezyang/b/pytorch/ff.py:5, code: return x + x
        add: f32[3] = torch.ops.aten.add.Tensor(arg0_1, arg1_1);  arg0_1 = arg1_1 = None
        return (add,)
```

After (only `arg0_1` and no more `arg1_1`):
```
 ====== Forward graph 0 ======
 <eval_with_key>.4 class <lambda>(torch.nn.Module):
    def forward(self, arg0_1: f32[3]):
        # File: dynamo/test_dup_global.py:8, code: return x + x
        add: f32[3] = torch.ops.aten.add.Tensor(arg0_1, arg0_1);  arg0_1 = None
        return (add,)
```

#### FSDP Test
Before we error on
```
File "/.../pytorch/torch/_guards.py", line 244, in __post_init__
    assert self.input_source_a != self.input_source_b
```
and now there is no error.

---
The rename from `name_to_input` to `input_name_to_proxy` is not part of the core logic change and is a remnant from initial attempts. I can undo it later if desired, but I also feel that the new name is more informative. It also fixes the type annotation.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Comment by awgu** [comment]
@ezyang Could I get a quick review to see if the approach makes sense?
1. I could not patch your PRs with `git apply` for some reason, so I manually patched them (hence the two PRs below this one).
2. The distributed test failure is because I used `inductor` backend without guarding the test with the appropriate decorators, so CI fails due to incompatibility. Locally, the test passes. I will fix this in the next time I push.
3. If you have any suggestions for how to better unit test this, I am all ears.
4. I did not dig into `speculate_subgraph()`. I am not sure if I would need to augment the logic there.

File: torch/_dynamo/output_graph.py [hunk 1]
```diff
@@ -358,20 +363,21 @@ def get_submodule(self, keys):
...
+                if candidate_name not in self.input_name_to_proxy:
+                    name = candidate_name
                     break
 
-        if self.name_to_input:
-            prev_name = next(reversed(self.name_to_input))
-            ctx = self.graph.inserting_after(self.name_to_input[prev_name])
+        if self.input_name_to_proxy:
+            prev_name = next(reversed(self.input_name_to_proxy))
+            ctx = self.graph.inserting_after(self.input_name_to_proxy[prev_name].node)
```
Comment (awgu) [inline torch/_dynamo/output_graph.py:375]: This is just a rename from `name_to_input` to `input_name_to_proxy` and `dict` value change from `fx.Node` to `fx.Proxy` from my initial attempt. I can undo this if desired.

File: torch/_dynamo/variables/torch.py [hunk 2]
```diff
@@ -836,10 +836,11 @@ def speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
             # Setup the subgraph we're going to capture into
             tx.output.graph = torch.fx.Graph()
             tx.output.graphargs = []
-            tx.output.name_to_input.clear()
+            tx.output.input_name_to_proxy.clear()
 
             args = []
             # One argument to graph per sub_args
+            # TODO (awgu): Do we need to de-dup here?
```
Comment (awgu) [inline torch/_dynamo/variables/torch.py:843]: This comment is not meant to be landed. I need to figure out if we need to change the logic here.

File: torch/_dynamo/output_graph.py [hunk 3]
```diff
@@ -265,11 +269,11 @@ def __init__(
         self.random_values_var = None
         self.initial_random_state = ()
         self.unspec_variable_map: Dict[str, UnspecializedPythonVariable] = {}
-        # Enables creating unique node names by tracking
-        # all current placeholder node names
-        self.name_to_input: OrderedDict[
-            str, Optional[fx.Proxy]
```
Comment (awgu) [inline torch/_dynamo/output_graph.py:271]: Previously, I think the typing should be `OrderedDict[str, fx.Node]`. If we choose to not include the refactor of `name_to_input` -> `input_name_to_proxy`, then we can still fix the type annotation.

File: torch/_dynamo/variables/builder.py [hunk 4]
```diff
@@ -584,6 +584,9 @@ def tensor_should_specialize(self):
         )
 
     def wrap_sym(self, value: Union[torch.SymInt, torch.SymFloat]):
+        is_duplicate_sym = self.get_source() in self.tx.output.input_source_to_var
```
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:587]: nit: 

```
if r := self.tx.output.input_source_to_var.get(self.get_source())
    return r
```

**Comment by ezyang** [comment]
> Note that create_graph_input() and add_grapharg() are not 1:1. For a constant source, TorchDynamo still calls create_graph_input() but not add_grapharg().

@voznesenskym can you explain this? This seems pretty strange to me

**Comment by ezyang** [comment]
> If anyone has any suggestion on how to coalesce this logic and avoid having to remember to update input_source_to_var for each create_graph_input(), I would love to hear it.

Looking over the code structure, it does look like you're hosed without a refactor. This strategy seems fine to start.

File: torch/_dynamo/variables/torch.py (hunk 2 above)
Comment (ezyang) [reply torch/_dynamo/variables/torch.py:843]: These arguments all represent locals, so no, I don't think they need to be deduped. The relevant case is

```
torch.cond(pred, true_fn, false_fn, [x, x])
```

you could try to test it.

**Review by ezyang — APPROVED** [review]
This looks good; enforcing the invariant more uniformly probably requires more refactor.

**Review by voznesenskym — APPROVED** [review]

**Pull request closed without merging**